### PR TITLE
Add cluster-proportional-vertical-autoscaler as a seperate deployment for ksm

### PR DESCRIPTION
**Pull request by benjaminhuo**
Add cluster-proportional-vertical-autoscaler as a seperate deployment for kube-state-metrics because the original addon resizer was removed

### CODE REVIEW

**Comment by benjaminhuo** [comment]
please hold this pr for now, I need more tests

Thanks
Ben

**Comment by benjaminhuo** [comment]
I made some modification, it works now.
In my testing env with 3 nodes, I've to increase the baseMemory to 220, with cpu step 1m, memory step 2Mi to avoid the oom of ksm.
So seems that we've to adjust the basememory and the cpu/mem steps

**Comment by benjaminhuo** [comment]
Sure

**Comment by benjaminhuo** [comment]
@brancz I've update my PR , would you take a look again?

File: jsonnet/kube-prometheus/kube-prometheus-ksm-autoscaler.libsonnet [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+      local c = [kubeStateMetricsAutoscaler];
+  
+      deployment.new('kube-state-metrics-autoscaler', 1, c, podLabels) +
+      deployment.mixin.metadata.withNamespace($._config.namespace) +
+      deployment.mixin.metadata.withLabels(podLabels) +
+      deployment.mixin.spec.selector.withMatchLabels(podLabels) +
+      deployment.mixin.spec.template.spec.withNodeSelector({ 'kubernetes.io/os': 'linux' }) +
+      deployment.mixin.spec.template.spec.securityContext.withRunAsNonRoot(true) +
+      deployment.mixin.spec.template.spec.securityContext.withRunAsUser(65534) +
+      deployment.mixin.spec.template.spec.withServiceAccountName('kube-state-metrics'),
```
Comment (brancz) [inline jsonnet/kube-prometheus/kube-prometheus-ksm-autoscaler.libsonnet]: This doesn't need all the permissions kube-state-metrics needs does it? I think this should have a distinct ServiceAccount with only the permissions it needs.

Comment (benjaminhuo) [reply jsonnet/kube-prometheus/kube-prometheus-ksm-autoscaler.libsonnet]: It needs patch permission of kube-state-metrics. I'll try to put it in a seperate service account

**Comment by benjaminhuo** [comment]
@brancz I've updated my PR, would you help to review it?
I'm not sure why the CI fails

**Comment by brancz** [comment]
Sorry CI failing is currently our fault. #264 is trying to fix this. Sorry for the invonvenience.

**Comment by benjaminhuo** [comment]
@brancz  I've pushed again, and the CI works now :)

File: examples/kustomize.jsonnet [hunk 2]
```diff
@@ -6,14 +6,15 @@ local kp =
...
+  //{ ['ksm-autoscaler-' + name]: kp.ksmAutoscaler[name] for name in std.objectFields(kp.ksmAutoscaler) } +
   { ['00namespace-' + name]: kp.kubePrometheus[name] for name in std.objectFields(kp.kubePrometheus) } +
   { ['0prometheus-operator-' + name]: kp.prometheusOperator[name] for name in std.objectFields(kp.prometheusOperator) } +
   { ['node-exporter-' + name]: kp.nodeExporter[name] for name in std.objectFields(kp.nodeExporter) } +
   { ['kube-state-metrics-' + name]: kp.kubeStateMetrics[name] for name in std.objectFields(kp.kubeStateMetrics) } +
   { ['alertmanager-' + name]: kp.alertmanager[name] for name in std.objectFields(kp.alertmanager) } +
   { ['prometheus-' + name]: kp.prometheus[name] for name in std.objectFields(kp.prometheus) } +
   { ['prometheus-adapter-' + name]: kp.prometheusAdapter[name] for name in std.objectFields(kp.prometheusAdapter) } +
-  { ['grafana-' + name]: kp.grafana[name] for name in std.objectFields(kp.grafana) };
+  { ['grafana-' + name]: kp.grafana[name] for name in std.objectFields(kp.grafana) };
```
Comment (brancz) [inline examples/kustomize.jsonnet]: this seems like an unnecessary whitespace change

Comment (benjaminhuo) [reply examples/kustomize.jsonnet]: > this seems like an unnecessary whitespace change

Fixed

**Comment by benjaminhuo** [comment]
> As we are not using this at all, @benjaminhuo do you think you could take on maintainership for this? Generally this looks ok to me otherwise.
> 
> What do others think @s-urbaniak @metalmatze @LiliC ?

I'd like to become maintainer of this , but some kind of guidance from you all will be great :)

File: examples/kustomize.jsonnet [hunk 3]
```diff
@@ -6,6 +6,7 @@ local kp =
   };
 
 local manifests =
+  //{ ['ksm-autoscaler-' + name]: kp.ksmAutoscaler[name] for name in std.objectFields(kp.ksmAutoscaler) } +
```
Comment (LiliC) [inline examples/kustomize.jsonnet]: Can we just add a comment explaining what this enables and when a user would need it?

Comment (benjaminhuo) [reply examples/kustomize.jsonnet]: > Can we just add a comment explaining what this enables and when a user would need it?

Sure

**Comment by brancz** [comment]
Maintainership for this would just entail that when there are problems with this we will ping you, and give you merge access to the repo :)

**Comment by benjaminhuo** [comment]
Thanks for the invitation. I'm happy to accept that

Ben
